slang.h

```diff
@@ -810,6 +829,51 @@ extern "C"
 
     #define SLANG_UUID_ISlangFileSystemExt { 0x5fb632d2, 0x979d, 0x4481, { 0x9f, 0xee, 0x66, 0x3c, 0x3f, 0x14, 0x49, 0xe1 } }
 
+    /* Identifies different types of writer target*/
+    typedef unsigned int SlangWriterChannel;
+    enum
+    {
+        SLANG_WRITER_CHANNEL_DIAGNOSTIC,
+        SLANG_WRITER_CHANNEL_STD_OUTPUT,
+        SLANG_WRITER_CHANNEL_STD_ERROR,
+        SLANG_WRITER_CHANNEL_COUNT_OF,
+    };
+
+    typedef unsigned int SlangWriterMode;
+    enum 
+    {
+        SLANG_WRITER_MODE_TEXT,
+        SLANG_WRITER_MODE_BINARY,
+    };
+
+    /** A stream typically of text, used for outputting diagnostic as well as other information.
+    */
+    struct ISlangWriter : public ISlangUnknown
+    {
+    public:
+            /** Write the formatted string with the used args. If returns SLANG_E_NOT_IMPLEMENTED, will use a default internal implementation, and call write with result
+            @param format Format string
+            @param args The arguments
+            @return SLANG_E_NOT_IMPLEMENTED if not implemented, SLANG_OK on success */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL writeVaList(const char* format, va_list args) = 0;
+            /** Write text to the writer
+            @param chars The characters to write out
+            @param numChars The amount of characters
+            @returns SLANG_OK on success */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL write(const char* chars, size_t numChars) = 0;
+            /** Flushes any content to the ouput */
+        virtual SLANG_NO_THROW void SLANG_MCALL flush() = 0;
+            /** Determines if the writer stream is to the console, and can be used to alter the output 
+            @returns Returns true if is a console writer */
+        virtual SLANG_NO_THROW SlangBool SLANG_MCALL isConsole() = 0;
+            /** Set the mode for the writer to use
+            @param mode The mode to use
+            @returns SLANG_OK on success */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL setMode(SlangWriterMode mode) = 0;
```

Comment:
The problem comes from we can ask for binary output and it be stdout. That internally to slang if it stdout is detected as not being 'console' it writes the binary data out. BUT when it does this on windows, it has to make the stream binary, to stop windows doing \n -> \r\n expansion and therefore mess up the binary.
I'm not seeing an easy way to fix this from the 'outside' - because when the stream is created, it would be non trivial to know what that later it would necessarily output to binary or text.

Reply:
As per some of our side-band conversations, I think the best approach here will be to stop having the tests work in terms of wrapping `slangc`'s command-line behavior (intercepting I/O, etc.) and instead just have a test fixture to drive the API more directly. All of the logic that creates problems here (detecting a console, setting stdout to binary) is specific to when Slang is running in its distinguished "command-line mode" which only `slangc` should ever be setting.